[CORDA-2486] Improve transaction deserialisation errors and fix possible migration issue

JamesHR3 opened the pull request:
This PR covers two issues, one as described in CORDA-2486, and another I encountered while trying to reproduce it.

The first issue is that an error is seen when querying vault states if the finance contracts cordapp is upgraded such that an extra field is added to the Cash contract state. The error is only seen if the field is non-nullable. The reason this happens is that the state has evolved in way that is non backwards-compatible, which is detected by the serialisation framework. This error should occur, but was being reported in a way that was not totally clear. This change adds a new exception type to report these problems better, and marks some RPC operations as throwing this exception so that the error message is reported.

The second issue is that an error is seen if the persistent Cash state is migrated to a new version, only the contracts jar is upgraded, and an attempt is made to issue cash with the new state. This occurs because the database migration scripts are bundled with the corda-finance-workflow jar, but the persistent states are defined in the corda-finance-contracts jar. This change moves the migration scripts to be packaged with the contracts jar.

Changes:
 - Move the finance migration scripts to the contracts jar.
 - Add a new exception to better report errors during transaction deserialisation
 - Update the RPC ops methods with `@Throws` annotations to describe what exceptions are thrown. This leads to the error messages from these exceptions being reported at the shell (is this desirable?)

tlil commented on core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt:
```diff
@@ -11,6 +11,7 @@ import net.corda.core.flows.StateMachineRunId
 import net.corda.core.identity.AbstractParty
 import net.corda.core.identity.CordaX500Name
 import net.corda.core.identity.Party
+import net.corda.core.internal.TransactionDeserialisationException
```
`TransactionDeserialisationException` is in `net.corda.core.internal`. This is now getting exposed by `CordaRPCOps` which is in `net.corda.core.messaging`. Wouldn't we want to avoid exposing the `internal` package here? The latter forms part of public API.

JamesHR3 replied:
Ah good point, that shouldn't happen. It's an exception I added, so either I can move it into a different package (such that it's still part of the public API), or detect when it is thrown in the vault query code and throw something else (perhaps the `VaultQueryException`) instead. I'll take a proper look tomorrow.

josecoll replied:
@JamesHR3 the transaction deserialization error should bubble up into the public  VaultQueryException (eg. not made public and thrown at the Public API level as it is a lower level failure condition - not directly related to Vault Queries)

tudor-malene commented:
It's not illogical at all.

The changelog files should only live in the workflows jar, and as a matter of fact so should the PersistentStates (but for historical reasons we can't move them).

It is a concern of the node what it can store locally and what field it understands.

Normally database changes can't happen at runtime when a node downloads a random "kernel" jar from a peer.

It's only when you update your flows to understand the new fields and the rpc clients.

JamesHR3 commented:
That does make sense. Why can't the persistent states be moved to workflows? Is it just that when finance was split into different jars there wasn't a sensible way to carry out the split such that that happens? Or is it something more fundamental? (I'm not completely familiar with the finance code).

tudor-malene commented:
> That does make sense. Why can't the persistent states be moved to workflows? Is it just that when finance was split into different jars there wasn't a sensible way to carry out the split such that that happens? Or is it something more fundamental? (I'm not completely familiar with the finance code).

The original design was that states that need to be saved to the database should extend from `QueryableState`. Which in turn returns the `PersistentState`s.

I lobbied to remediate this before C4, but with no luck. Maybe C5.

JamesHR3 commented:
Ah ok. I'll revert that change.

josecoll commented:
State definitions and their persistent representation are closely coupled by design (eg. one is the in-memory data structure representation, the other its persistent representation).
A State's persistent storage is further decoupled from how Flows manipulate / process state fields.
The `queryable` interface was introduced to make persistence optional (remember, states are stored as binary blobs on the ledger anyway), and at a time before we adopted Hibernate. 
In an ideal world we would define the State once (and annotate its fields with persistent annotations) without the need for an additional `Queryable` interface implementation.

JamesHR3 commented:
I've done the following:
 - Moved the migration scripts back to the workflow jar
 - Ensured that `_queryBy` in the `NodeVaultService` throws a `VaultQueryException`
 - Allowed for reporting the cause of a `VaultQueryException`
 - Removed the `@Throws` annotations. I found that with it the logs were not giving any useful information about what had gone wrong. The console contained the exception error message, which was often quite verbose in the case of serialisation errors. I've therefore reverted to not having them (which prints an error log to the console and gives the full stack trace in the logs).

tudor-malene commented on core/src/main/kotlin/net/corda/core/internal/TransactionUtils.kt:
```diff
@@ -79,11 +79,17 @@ fun <T : Any> deserialiseComponentGroup(componentGroups: List<ComponentGroup>,
+ * Exception raised if an error was encountered while attempting to deserialise a component group in a transaction.
+ */
+class TransactionDeserialisationException(groupEnum: ComponentGroupEnum, index: Int, cause: Exception):
+        Exception("Failed to deserialise group $groupEnum at index $index in transaction: ${cause.message}", cause)
```
Is this message correct?
Is the transactionId in the exception message?

JamesHR3 replied:
Good question. Annoyingly, the `deserialiseComponentGroup` function (where this is thrown from) doesn't have the transactionId passed to it, and I wanted to avoid introducing a lot of plumbing to get this through into the message. (The original message was very similar but didn't have the message from the exception cause in it.)
